### PR TITLE
add support for lambda expressions

### PR DESCRIPTION
**Pull request by HliasMpGH**
FIRST PART OF ISSUE #5

This feature allows the use of lambda expressions in CodeBlocks and within MethodSpecs.

A usage of the feature can be:

```
// build a method that uses lambdas
MethodSpec method = MethodSpec.methodBuilder("method")
    .addCode("methodCall(")
    .addLambda("5 + 7") // lambda with no inputs and (String) body
    .addCode(", ")
    .addLambda("method1(); method2();") // lambda with multiple statements
    .addCode(");\n")
    .build();
```
**to produce:**
```
void method() {
    methodCall(() -> 5 + 7, () -> {method1(); method2();});
}
```
The visibility of the input types in the result is a conditional that can be configured as follows:
```
// declare the inputs of the function
ParameterSpec p1 = ParameterSpec.builder(TypeName.INT, "x").build();
ParameterSpec p2 = ParameterSpec.builder(TypeName.DOUBLE, "y").build();

...
.addLambda(List.of(p1, p2), true, "5 + 7")
...
```
**that produces**:

```
(int x, double y) -> 5 + 7
```

or:
```
...
addLambda(List.of(p1, p2), "5 + 7")
...
```
**to get**:

```
(x, y) -> 5 + 7
```

It is not required to specify if the body of the lambda is an expression or not when using the feature. The output will reflect, depending on the inputs, the wanted and most importantly, valid, result.

if multiple statements are given as a body to the lambda, in the ouput it will be included in braces.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+target
+.vscode
```
Comment (Randgalt) [inline .gitignore:2]: Please make this a separate commit

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 2]
```diff
@@ -415,6 +418,126 @@ public Builder add(CodeBlock codeBlock) {
...
+     * Should be used with {@link #add(String, Object...) addCode},
+     * to provide specific behaviour such as
+     * <b>methodcall((int x, int y) -> x + y, 5)</b>.
+     * @param parameters the input parameters of the function.
+     * @param emitTypes true if the types of the inputs should
+     * be emitted on the result.
+     * @param body the body of the function.
+     */
+    public Builder addLambda(Iterable<ParameterSpec> parameters, boolean emitTypes, CodeBlock body) {
+      int paramsLen = 0;
```
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/CodeBlock.java:432]: I'd move these checks into the processing of the parameter stream below. This way we don't go over the parameters list twice. It's small but no reason to double iterate the list.

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 3]
```diff
@@ -415,6 +418,126 @@ public Builder add(CodeBlock codeBlock) {
...
+     * Structures a lambda function based on some inputs and a CodeBlock body.<br>
+     * Should be used with {@link #add(String, Object...) addCode},
+     * to provide specific behaviour such as
+     * <b>methodcall((int x, int y) -> x + y, 5)</b>.
+     * @param parameters the input parameters of the function.
+     * @param emitTypes true if the types of the inputs should
+     * be emitted on the result.
+     * @param body the body of the function.
+     */
+    public Builder addLambda(Iterable<ParameterSpec> parameters, boolean emitTypes, CodeBlock body) {
```
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/CodeBlock.java:431]: wdyt of making `emitTypes` an enum instead. Something like `public enum LambdaMode ...`. This is a better U/X for users and allows us to add more modes should we ever need to.

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 4]
```diff
@@ -415,6 +418,126 @@ public Builder add(CodeBlock codeBlock) {
...
+      .collect(Collectors.joining(", "));
+
+      // on 0 or more than 1 inputs, or in case of type emission,
+      // parentheses are mandatory
+      if (paramsLen == 0 || paramsLen > 1 || emitTypes) {
+        inputSide = "(" + inputSide + ")";
+      }
+
+      // the body of the lambda (right side)
+      String bodySide = body.toString().replaceAll("\n", "");
```
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/CodeBlock.java:457]: This seems wrong. Why are we doing this? We should not reformat code that has already been specified by users.

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 5]
```diff
@@ -415,6 +418,126 @@ public Builder add(CodeBlock codeBlock) {
...
+
+      // in case of multiple statements, braces are mandatory
+      if (bodySide.contains(";")) {
+        bodySide = "{" + bodySide + "}";
```
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/CodeBlock.java:461]: This is forcing a coding style. Users need to be able to specify newlines, indents, etc. Many people will not like the output of this.

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 6]
```diff
@@ -415,6 +418,126 @@ public Builder add(CodeBlock codeBlock) {
...
+      }
+
+      // the full lambda structure
+      add(inputSide + " -> " + bodySide);
```
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/CodeBlock.java:465]: Assuming spaces around `->` is wrong. Like all things, we must allow users to specify how code is formatted. FYI - this is why I suggested a separate `LambdaBlock` class. There are simply too many options to force into method arguments.

File: pom.xml [hunk 7]
```diff
@@ -135,16 +135,6 @@
     <build>
         <pluginManagement>
             <plugins>
-                <plugin>
```
Comment (Randgalt) [inline pom.xml:138]: Please remove the 2 Javadoc commits as they negate each other.

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 3 above)
Comment (HliasMpGH) [reply src/main/java/com/squareup/javapoet/CodeBlock.java:431]: Thats sounds like a good improvement and thank your for your comment! Are you thinking something like enum {showTypes;} for type visibility and we can add more values if we need later?

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 6 above)
Comment (HliasMpGH) [reply src/main/java/com/squareup/javapoet/CodeBlock.java:465]: My idea was based on the rest of the codebase that seems to follow that kind of pattern. We have methods such as beginControlFlow() for example that does indeed emitt spaces before { to produce the output and doesnt allow user configuration. I take that this is coded that way because spaces before { on if statements and for loops, are a standard for java code.

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 4 above)
Comment (HliasMpGH) [reply src/main/java/com/squareup/javapoet/CodeBlock.java:457]: wdyt about having a new line after the { so for body inputs like:

```
int x = 6;
int y = 7;
return x + y;
```
we get outputs like:

```
input -> {
int x = 6;
int y = 7;
return x + y;
}
```
(or with proper indents if you think thats better).
By leaving the formatted user input as is, we might fall into outputs like:
```
input -> {int x =6;
int y = 7;
return x + y;
}
```
wdyt?

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 2 above)
Comment (HliasMpGH) [reply src/main/java/com/squareup/javapoet/CodeBlock.java]: I pushed latest. If you agree with the approach i can flag it as resolved, thanks.

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 3 above)
Comment (HliasMpGH) [reply src/main/java/com/squareup/javapoet/CodeBlock.java]: I pushed latest, the logic stays pretty much the same, but now we can add various methods if we need to in the future. When it is needed, we just add a mode to LambdaModes enum and add the case to the parent addLambda method (in CodeBlock). I added 2 new overloads for when the user defines a lambda with no inputs (so he can define a mode, or not and the default one is used) in case a new mode is added in the future that can be applied to said lambdas.

**Comment by Randgalt** [comment]
@HliasMpGH is there anyone else you know who can help review this?

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 4 above)
Comment (Randgalt) [reply src/main/java/com/squareup/javapoet/CodeBlock.java:458]: The braces placement should be customizable. No one will agree on where to put it. This is why I've been suggesting we have a `LambdaSpec` class. I don't see any way around it. If we had a `LambdaSpec` builder class we wouldn't need so many method overloads. This would also pay benefits for the `switch` PR. the body after the `->` arm of a switch is just like a lambda and a builder would help there. 

I'd like to see all these method overloads moved into a builder class, possibly with a bas class that can re-used for the switch PR.

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 5 above)
Comment (Randgalt) [reply src/main/java/com/squareup/javapoet/CodeBlock.java:462]: Indenting, newlines, etc. must all be configurable by users.

**Comment by HliasMpGH** [comment]
The implementation of the LambdaSpec class can be seen in #16. Closing
